File: bindings/flutter/lib/src/audio_session.dart
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'dart:async';
import 'dart:convert';
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import 'bindings/native_library.dart';
import 'error_capture.dart';
import 'job_runner.dart';
import 'model.dart';
import 'models/chat.dart';
import 'models/delta.dart';
import 'native_strings.dart';
import 'session_base.dart';

/// Options for an audio session.
class AudioSessionOptions {
  const AudioSessionOptions({this.language});
  final String? language;
  Map<String, Object?> toJson() => <String, Object?>{
        'type': 'audio',
        if (language != null) 'language': language,
      };
}

/// A batch or streaming speech-to-text session.
class AudioSession extends Session {
  AudioSession._(Model model, int handle) : super(model: model, handle: handle);

  factory AudioSession.create(Model model, AudioSessionOptions options) =>
      AudioSession._(model, Session.createHandle(model, options.toJson()));

  /// Transcribe a file or a base64-encoded blob in one shot.
  Future<TranscriptionResult> transcribeFile(
    String path, {
    String? language,
    bool translate = false,
  }) async {
    final request = <String, Object?>{
      'path': path,
      if (language != null) 'language': language,
      'translate': translate,
    };
    final result = await withCString<Future<Map<String, Object?>>>(
      jsonEncode(request),
      (ptr) => runSimpleJob(
        (completionPtr, userData, outJob) =>
            NativeLibrary.instance.bindings.flm_session_transcribe_async(
          handle,
          ptr,
          nullptr,
          completionPtr,
          userData,
          outJob,
        ),
      ),
    );
    return TranscriptionResult.fromJson(result);
  }

  /// Streaming transcription of live PCM audio.
  ///
  /// Call [pushAudio] to feed chunks; a final [pushAudio] with `isFinal: true`
  /// tells the model to emit the last segment and close.
  Stream<SpeechDelta> startStreaming({
    int sampleRate = 16000,
    int channels = 1,
    String? language,
  }) {
    final bindings = NativeLibrary.instance.bindings;
    final request = <String, Object?>{
      'streaming': true,
      'sample_rate': sampleRate,
      'channels': channels,
      if (language != null) 'language': language,
    };
    late StreamController<SpeechDelta> controller;
    JobHandles<SessionDelta>? handles;
    controller = StreamController<SpeechDelta>(
      onListen: () {
        withCString<void>(jsonEncode(request), (ptr) {
          handles = runStreamingJob(
            (deltaPtr, completionPtr, userData, outJob) =>
                bindings.flm_session_transcribe_async(
              handle,
              ptr,
              deltaPtr,
              completionPtr,
              userData,
              outJob,
            ),
          );
          handles!.stream.listen(
            (d) {
              if (d is SpeechDelta) controller.add(d);
            },
            onError: controller.addError,
            onDone: controller.close,
          );
          handles!.result.catchError((_) => <String, Object?>{});
        });
      },
      onCancel: () => handles?.cancel(),
    );
    return controller.stream;
  }

  /// Push a chunk of PCM audio into a live transcription session.
  void pushAudio(
    Uint8List pcm, {
    required int sampleRate,
    required int channels,
    bool isFinal = false,
  }) {
    final ptr = calloc<Uint8>(pcm.length);
    try {
      ptr.asTypedList(pcm.length).setAll(0, pcm);
      final status = NativeLibrary.instance.bindings.flm_session_push_audio(
        handle,
        ptr.cast<Void>(),
        pcm.length,
        sampleRate,
        channels,
        isFinal ? 1 : 0,
      );
      checkStatus(status, fallbackMessage: 'flm_session_push_audio failed');
    } finally {
      calloc.free(ptr);
    }
  }
}

File: bindings/ios/Sources/FoundryLocal/AudioSession.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import Foundation
import FoundryLocalMobile

/// Options for a speech-to-text session.
public struct AudioSessionOptions: Sendable {
    /// BCP-47 language tag, e.g. `"en"` or `"en-US"`. When `nil` the model auto-detects.
    public var language: String?
    public var temperature: Double?
    public var maxOutputTokens: Int?

    public init(language: String? = nil, temperature: Double? = nil, maxOutputTokens: Int? = nil) {
        self.language = language
        self.temperature = temperature
        self.maxOutputTokens = maxOutputTokens
    }

    func encodeAsJSON() -> String {
        var payload: [String: Any] = ["type": "audio"]
        if let language { payload["language"] = language }
        if let temperature { payload["temperature"] = temperature }
        if let maxOutputTokens { payload["max_output_tokens"] = maxOutputTokens }
        return payload.jsonString() ?? "{}"
    }
}

/// A speech-to-text session over a loaded speech model.
///
/// Two modes:
///   * **File**: pass a WAV/MP3 path or base64 blob to ``transcribeStreaming`` and get
///     partial + final segments back.
///   * **Streaming microphone**: start the session with
///     ``startStreamingTranscription``, then push PCM chunks via ``pushAudio``.
public final class AudioSession: @unchecked Sendable {
    public let handle: flm_session
    private let released = ManagedAtomicBool()

    init(model: Model, options: AudioSessionOptions) throws {
        var out: flm_session = 0
        let json = options.encodeAsJSON()
        let status = json.withCString { flm_session_create(model.handle, $0, &out) }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
        self.handle = out
    }

    deinit { close() }

    public func close() {
        if released.exchange(true) { return }
        _ = flm_session_release(handle)
    }

    // MARK: - File / blob transcription

    /// Transcribe an audio file at a filesystem path, streaming partial and final
    /// segments.
    public func transcribeStreaming(
        path: String,
        language: String? = nil,
        translate: Bool = false
    ) -> AsyncThrowingStream<SpeechDelta, any Error> {
        var payload: [String: Any] = ["path": path, "translate": translate]
        if let language { payload["language"] = language }
        return transcribeStream(payload: payload)
    }

    /// Transcribe raw PCM (or any format the runtime supports) given as base64 data.
    public func transcribeStreaming(
        base64: String,
        format: String,
        sampleRate: Int,
        channels: Int = 1,
        language: String? = nil,
        translate: Bool = false
    ) -> AsyncThrowingStream<SpeechDelta, any Error> {
        var payload: [String: Any] = [
            "data_base64": base64,
            "format": format,
            "sample_rate": sampleRate,
            "channels": channels,
            "translate": translate,
        ]
        if let language { payload["language"] = language }
        return transcribeStream(payload: payload)
    }

    /// Non-streaming file transcription. Await the full result.
    public func transcribe(path: String, language: String? = nil, translate: Bool = false) async throws -> TranscriptionResult {
        var payload: [String: Any] = ["path": path, "translate": translate]
        if let language { payload["language"] = language }
        let json = payload.jsonString() ?? "{}"
        return try await runAsyncJob(
            decode: { job in
                let text = try takeJobResultJSON(job)
                return try flmJSONDecoder.decode(TranscriptionResult.self, from: Data(text.utf8))
            },
            submit: { [handle] userData, _, onComplete, outJob in
                json.withCString { jsonPtr in
                    flm_session_transcribe_async(handle, jsonPtr, nil, onComplete, userData, outJob)
                }
            }
        )
    }

    // MARK: - Streaming microphone

    /// Start a live PCM transcription. Feed chunks with ``pushAudio``; partial and
    /// final segments arrive on the returned stream.
    ///
    /// Once you have sent the last chunk with `isFinal == true`, the stream will
    /// finish naturally after the model flushes its remaining segments.
    public func startStreamingTranscription(
        language: String? = nil,
        translate: Bool = false
    ) -> AsyncThrowingStream<SpeechDelta, any Error> {
        var payload: [String: Any] = ["streaming": true, "translate": translate]
        if let language { payload["language"] = language }
        return transcribeStream(payload: payload)
    }

    /// Push a chunk of PCM into a live transcription. Call with `isFinal: true` on
    /// the last chunk to flush the tail.
    public func pushAudio(
        _ data: Data,
        sampleRate: Int32,
        channels: Int32,
        isFinal: Bool
    ) throws {
        try data.withUnsafeBytes { rawBuffer in
            let base = rawBuffer.baseAddress
            let status = flm_session_push_audio(
                handle, base, data.count, sampleRate, channels, isFinal ? 1 : 0
            )
            if status != FLM_OK {
                throw FoundryLocalError.fromCurrent(status: status)
            }
        }
    }

    // MARK: - Internal

    private func transcribeStream(payload: [String: Any]) -> AsyncThrowingStream<SpeechDelta, any Error> {
        guard let json = payload.jsonString() else {
            return AsyncThrowingStream { $0.finish(throwing: FoundryLocalError(code: .invalidArgument, message: "failed to encode transcription request")) }
        }
        return runStreamingJob(
            decodeDelta: { SpeechDelta.from(cValue: $0) },
            submit: { [handle] userData, onDelta, onComplete, outJob in
                json.withCString { jsonPtr in
                    flm_session_transcribe_async(handle, jsonPtr, onDelta, onComplete, userData, outJob)
                }
            }
        )
    }
}
